### ISSUE TITLE
Embedding服务性能优化

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#169]
## 使用 ONNX Runtime作为推理引擎

常用的有 SentenceTransformer 和 ONNXRuntime，ONNX Runtime 在 CPU 场景性能更好。

ONNX Runtime 通过静态图 + Kernel Fusion + SIMD 优化，大幅减少 Python 调度与内存开销。

详细对比原因如下表：

| 对比项 | Sentence Transformers(PyTorch) | ONNX Runtime |
| ------- | -------------------------------- | ---------------- |
| 模型执行方式 | 动态计算图 | 静态优化图 |
| 调度机制 | Python 解释执行 | C++优化执行 |
| CPU 优化 | 一般 | 高度优化（AVX/BLAS/ThreadPool） |

doing

## 量化模型

CPU 场景下，首先应该考虑的就是模型量化。

我们拿 bge-m3 举例：默认情况下，其 dtype 是 FP32，CPU 推理极慢。我们可以通过动态量化转换成 int8 或 int8+fp16 混合精度模型。

## OpenVINO

Open Visual Inference and Neural Network Optimization，是 Intel 开发的推理加速框架。加速措施包括在底层做算子融合、内存布局优化、图编译优化、多线程推理引擎（OpenMP）、内核加速等。

重要的是：👉 在 Intel CPU 上，它可以充分利用 AVX2/AVX512/VNNI 指令集 来加速矩阵乘法。

在 onnxruntime 中，将 inference session的 peoviders 数组头部添加 OpenVINOExecutionProvider，并安装 onnxruntime-openvino pip包即可。

```

import onnxruntime as ort

session = ort.InferenceSession(
    "model.int8.onnx",
    providers=["OpenVINOExecutionProvider", "CPUExecutionProvider"]
)
```

在 BERT/BGE/Reranker 这种 Transformer 模型 上，通常带来：
- 1.3~2.5 倍加速
- 内存减少 20~40%

## Batch批处理

Embedding 服务面对的是 “高并发、小请求” 的典型场景。

如果不做 batching，每个请求都是独立的：

`-> tokenizer -> inference -> normalize -> response`

那意味着每条请求都要：
- 重新构造 ONNX session 输入；
- 启动一次矩阵乘法；
- 写出结果；
- 清理缓存。

CPU 会出现大量 “碎片化推理”，性能极差。

而批处理可以：
- 将几毫秒内的多个请求合并；
- 一次性跑完；
- 再拆结果返回。

这就是所谓的 dynamic batching，几乎所有的高性能推理引擎（如 Triton Inference Server、vLLM、TEI）都用这个机制。

## Python ThreadPoolExecutor

Python 有 GIL（全局解释器锁），所以同一时刻只有一个线程执行 Python 字节码，可以通过 event loop 、ThreadPool Executor 减轻其影响。但对于 CPU 密集型任务（如矩阵乘法、模型计算），ThreadPoolExecutor 无法显著加速；但对于 IO 密集型任务（如 HTTP请求、数据库操作、ONNX 推理）非常有效，因为这些任务在等待 IO 时会释放 GIL。而 ONNX Runtime 属于 CC++ 原生库执行的计算任务，其内部会释放 GIL，因此 ThreadPoolExecutter 可以并行执行多个推理任务。
